### PR TITLE
improvement(restore): add script for restoring from backup file w/o having cluster

### PR DESCRIPTION
**Pull request by weizhe0422**
[Longhorn 1521](https://github.com/longhorn/longhorn/issues/1521)

Signed-off-by: Ray Chang <ray.chang@suse.com>

### CODE REVIEW

File: examples/restore_to_file.Dockerfile.template [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM rancher/longhorn-engine:v0.4.1
```
Comment (innobead) [inline examples/restore_to_file.Dockerfile.template:1]: Please use `longhornio/longhorn-engine` image instead, also I would suggest using 1.3.1 image.  We also need to update `restore_to_file.yaml.template` as well.

**Review by innobead — CHANGES_REQUESTED** [review]
Suggest createing a script `restore-backup-to-file.sh` in scripts folder and use docker command inside the script to run `longhornio/longhorn-engine` to exeucte instead of having Dockerfile. 

A executable script is better than a Dockerfile which means users need to build first, then run later..

File: examples/restore_to_file.Dockerfile.template (hunk 1 above)
Comment (weizhe0422) [reply examples/restore_to_file.Dockerfile.template]: Update the Longhorn-engine image version as 1.3.1

https://github.com/longhorn/longhorn/blob/2a1c04cf7ea3696c0552e986b64fab5f1b13fd69/examples/restore_to_file.yaml.template#L19

**Comment by weizhe0422** [comment]
I will modify the DOCKERFILE generated by shell script.

**Review by innobead — CHANGES_REQUESTED** [review]
@weizhe0422 you misunderstood my previous review... I did not ask you to create a script to generate Dockerfile but rather run `docker run` command directly via `longhornio/longhorn-engine` to restore the backup to a file. 

for example:
```shell
# Use -e to add required environment variables
# Use -v to bind mount to specify the generate the place
docker run longhornio/longhorn-engine:v1.3.1 longhorn backup restore-to-file <backup-url> --backing-file <backing-file> --output-file <output-file> --output-format <output-format>
```

Also, actually, the backup could be based on a backing image, so need to consider this case as well. Please check `--backing-file`.

**Comment by weizhe0422** [comment]
I will reprocess the PR, sorry to misunderstand the comment.

**Comment by weizhe0422** [comment]
Script that starts the `longhornio/longhorn-engine` docker container to execute.

- w/o specify backing image
image
- w/ specify backing image
image

**Review by innobead — CHANGES_REQUESTED** [review]
Just noticed you changed the script with an interactive way, which is good for one time execution, but not good for automation, so it will limit how users can use it. 

I would suggest making the script to run directly with optional options, so it will let users to decide how they will use this script. WDYT?

You can refer to other scripts in scripts folder.

**Comment by weizhe0422** [comment]
I would get these parameters as script options instead of interactive mode.

**Comment by innobead** [comment]
> I would get these parameters as script options instead of interactive mode.

@weizhe0422 if it's ready to review, you can request to avoid it pending for a while. Thanks.

**Comment by weizhe0422** [comment]
Extract parameter as script options, and also check whether the required parameter is blank or not. 

- w/o specify backing image
image
- w specify backing image
image

File: scripts/restore-backup-to-file.sh [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+test -z ${aws_access_key} && ${aws_secret_access_key} && ${backup_url} && ${target_directory} && ${output_format} &&
+    echo "There are REQUIRED parameters you missed!" && exit 0
+
+if [ ${backing_file} == ""]
+then 
+   docker run -v /tmp/restore:/tmp/restore \
+              -e AWS_ACCESS_KEY_ID=${aws_access_key} -e AWS_SECRET_ACCESS_KEY=${aws_secret_access_key} \
+              longhornio/longhorn-engine:v1.3.1 longhorn backup \
+              restore-to-file ${backup_url} \
+              --output-file ${target_directory} --output-format ${output_format}
```
Comment (c3y1huang) [inline scripts/restore-backup-to-file.sh:79]: Can the user access the output file when the `${target_directory}` is not in `/tmp/restore`?

File: scripts/restore-backup-to-file.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+fi
+
+test -z ${aws_access_key} && ${aws_secret_access_key} && ${backup_url} && ${target_directory} && ${output_format} &&
+    echo "There are REQUIRED parameters you missed!" && exit 0
+
+if [ ${backing_file} == ""]
+then 
+   docker run -v /tmp/restore:/tmp/restore \
+              -e AWS_ACCESS_KEY_ID=${aws_access_key} -e AWS_SECRET_ACCESS_KEY=${aws_secret_access_key} \
+              longhornio/longhorn-engine:v1.3.1 longhorn backup \
```
Comment (c3y1huang) [inline scripts/restore-backup-to-file.sh:77]: What about making the version tag a parameter? Do you think it will be beneficial?

Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh:77]: I think this can be an optional parameter. If the user leaves the version number blank, "latest" is used instead.

Comment (innobead) [reply scripts/restore-backup-to-file.sh:77]: We can't use master-head in the script for users to use, because it's not dev image.

I think Chinya suggestion probably better to have a version parameter provided by users, but actually you can simply guide user to use the right version via doc like `./scripts/restore-backup-to-file.sh --version <version> ...` and <version> can be replaced by `v{{< current-version >}}` in the doc.

Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh:77]: I got it, I will just provide another argument for user to specify version and mentions it on document.

File: scripts/restore-backup-to-file.sh (hunk 2 above)
Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh:79]: I try to set the target directory so that it is not under `/tmp/restore`, it returns an error saying that the folder cannot be created or accessed.  And the template yaml file also mention it should be in `/tmp/restore`
https://github.com/longhorn/longhorn/blob/4ba39a989b4b482d51fd4bc651f61f2b419428bd/examples/restore_to_file.yaml.template#L23-L25

I will modify the script file to make `/tmp/restore` as the target directory prefix like `--output-file /tmp/restore/${target_directory}`

File: examples/restore_to_file.yaml.template [hunk 4]
```diff
@@ -16,7 +16,7 @@ spec:
       --output-file '/tmp/restore/<OUTPUT_FILE>'
       --output-format <OUTPUT_FORMAT>
     # the version of longhorn engine should be v0.4.1 or higher
-    image: rancher/longhorn-engine:v0.4.1
+    image: longhornio/longhorn-engine:v<LONGHORN_VERSION>
```
Comment (derekbit) [inline examples/restore_to_file.yaml.template:19]:
```suggestion
    image: longhornio/longhorn-engine:<LONGHORN_VERSION>
```

File: scripts/restore-backup-to-file.sh [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+   echo "          --target-directory /tmp/restore/volume.raw \ "
+   echo "          --output-format raw \ "
+   echo "          [--backing-file <backing_file_path>]"
+   echo ""
+   echo "      --aws-access-key         (Required) AWS credentials access key, e.g. AKIAIOSFODNN7EXAMPLE"
+   echo "      --aws-secret-access-key  (Required) AWS credentials access secret key, e.g. wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
+   echo "  -u, --backup-url             (Required) Backups URL, e.g. s3://backupbucket@us-east-1/backupstore?backup=backup-bd326da2c4414b02&volume=volumeexamplename"
+   echo "  -o, --output-file            (Required) Specify the output file name, which is saved in the /tmp/restore directory by default, e.g. volume.raw"
+   echo "  -f, --output-format          (Required) Output file foramt, and the supported options are raw or qcow2."
+   echo "  -v, --version                (Required) Specify Lognhorn version to execute, e.g. 1.3.2"
```
Comment (derekbit) [inline scripts/restore-backup-to-file.sh:66]:
```suggestion
   echo "  -v, --version                (Required) Specify Lognhorn version to execute, e.g. v1.3.2"
```

File: scripts/restore-backup-to-file.sh [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+		;;
+		*)
+		echo "Error! invalid flag: ${key}"
+		help="true"
+		break
+		;;
+	esac
+done
+
+usage () {
```
Comment (derekbit) [inline scripts/restore-backup-to-file.sh:53]: Can you add a short description of the script's purpose?
e.g.
```
# gzip -h
Usage: gzip [OPTION]... [FILE]...
Compress or uncompress FILEs (by default, compress FILES in-place).

Mandatory arguments to long options are mandatory for short options too.

  -c, --stdout      write on standard output, keep original files unchanged
...
...
```

File: scripts/restore-backup-to-file.sh [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+   echo "          --backup-url s3://backupbucket@ap-northeast-1/backupstore?backup=<backup_name>&volume=<volume_name> \ "
+   echo "          --target-directory /tmp/restore/volume.raw \ "
+   echo "          --output-format raw \ "
+   echo "          [--backing-file <backing_file_path>]"
+   echo ""
+   echo "      --aws-access-key         (Required) AWS credentials access key, e.g. AKIAIOSFODNN7EXAMPLE"
+   echo "      --aws-secret-access-key  (Required) AWS credentials access secret key, e.g. wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
+   echo "  -u, --backup-url             (Required) Backups URL, e.g. s3://backupbucket@us-east-1/backupstore?backup=backup-bd326da2c4414b02&volume=volumeexamplename"
+   echo "  -o, --output-file            (Required) Specify the output file name, which is saved in the /tmp/restore directory by default, e.g. volume.raw"
+   echo "  -f, --output-format          (Required) Output file foramt, and the supported options are raw or qcow2."
```
Comment (derekbit) [inline scripts/restore-backup-to-file.sh:65]:
```suggestion
   echo "  -f, --output-format          (Required) Output file foramt. Supported format: raw or qcow2."
```

File: examples/restore_to_file.yaml.template (hunk 4 above)
Comment (weizhe0422) [reply examples/restore_to_file.yaml.template]: Thank you and I will follow the comment to fix it.

File: scripts/restore-backup-to-file.sh [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+done
+
+usage () {
+   echo "USAGE: $0 --aws-access-key <your_aws_access_key>  \ "
+   echo "          --aws-secret-access-key <your_aws_secret_access_key> \ "
+   echo "          --backup-url s3://backupbucket@ap-northeast-1/backupstore?backup=<backup_name>&volume=<volume_name> \ "
+   echo "          --target-directory /tmp/restore/volume.raw \ "
+   echo "          --output-format raw \ "
+   echo "          [--backing-file <backing_file_path>]"
+   echo "Restore a Longhorn backup to a raw image or a qcow2 image w/o system installed."
```
Comment (derekbit) [inline scripts/restore-backup-to-file.sh:60]:
```suggestion
   echo "          [--backing-file <backing_file_path>]"
   
   echo "Restore a Longhorn backup to a raw or qcow2 formatted image."
```

**Review by derekbit — APPROVED** [review]
In general, LGTM.

File: scripts/restore-backup-to-file.sh (hunk 8 above)
Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh]: fix: removed the redundant words in description

File: examples/restore_to_file.yaml.template [hunk 9]
```diff
@@ -16,7 +16,7 @@ spec:
       --output-file '/tmp/restore/<OUTPUT_FILE>'
       --output-format <OUTPUT_FORMAT>
     # the version of longhorn engine should be v0.4.1 or higher
-    image: rancher/longhorn-engine:v0.4.1
+    image: longhornio/longhorn-engine:<LONGHORN_VERSION>
```
Comment (innobead) [inline examples/restore_to_file.yaml.template:19]: Because of changing this, you need to update https://longhorn.io/docs/<each version>/advanced-resources/data-recovery/recover-without-system/ accordingly.

File: scripts/restore-backup-to-file.sh [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+		output_format="$2"
+		shift # past argument
+		shift # past value
+		;;
+		-b|--backing-file)
+		backing_file="$2"
+		shift # past argument
+		shift # past value
+		;;
+		-v|--version)
```
Comment (innobead) [inline scripts/restore-backup-to-file.sh:36]: Wrong indent.

File: scripts/restore-backup-to-file.sh [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+		version="$2"
+		shift # past argument
+		shift # past value
+		;;
+		-h|--help)
```
Comment (innobead) [inline scripts/restore-backup-to-file.sh:41]: Wrong indent.

**Review by innobead — CHANGES_REQUESTED** [review]

File: scripts/restore-backup-to-file.sh [hunk 12]
```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
+
+while [[ $# -gt 0 ]]; do
+	key="$1"
```
Comment (innobead) [inline scripts/restore-backup-to-file.sh:4]: I think you are using a tab instead of empty space. For shell script, please use 2 spaces instead of a tab. https://google.github.io/styleguide/shellguide.html#s5.1-indentation (you can check other scripts)

**Review by innobead — CHANGES_REQUESTED** [review]

File: scripts/restore-backup-to-file.sh (hunk 12 above)
Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh]: I referenced another script but forgot to replace tabs with spaces and it was completely fixed.

I would look at the coding style documentation you provided.

File: examples/restore_to_file.yaml.template (hunk 9 above)
Comment (weizhe0422) [reply examples/restore_to_file.yaml.template:19]: Yes, I will also update the documentation to let users know that the version number parameter needs to be specified.

https://github.com/longhorn/website/pull/580

Comment (innobead) [reply examples/restore_to_file.yaml.template:19]: Good!

File: scripts/restore-backup-to-file.sh [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+   echo "  -b, --backing-file           (Optional) If the backup is based on a backing file, it should provide the backing file, e.g. /tmp/bakingfile.qcow2"
+   echo "  -h, --help                   Usage message"
+}
+
+if [[ $help ]]; then
+    usage
+    exit 0
+fi
+
+if [ ! ${aws_access_key} ] || [ ! ${aws_secret_access_key} ] ||
```
[the remark this answers is not in the document]
Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh:77]: After reference the [this](https://www.linux.org/threads/what-does-z-flag-mean-in-shell-script-used-in-if-condition.17899/post-54067) and also check `test help`, I think using `-z` is the formal way to check the string is empty or not.

[the remark this answers is not in the document]
Comment (derekbit) [reply scripts/restore-backup-to-file.sh]: > Usually, use -z to check whether the variable is an empty string or not, but think the bang operator could work as well.

Yeah.

File: scripts/restore-backup-to-file.sh [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+              longhornio/longhorn-engine:${version} longhorn backup \
+              restore-to-file ${backup_url} \
+              --output-file /tmp/restore/${output_file} --output-format ${output_format}
+else
+   docker run -v /tmp/restore:/tmp/restore \
+              -e AWS_ACCESS_KEY_ID=${aws_access_key} -e AWS_SECRET_ACCESS_KEY=${aws_secret_access_key} \
+              longhornio/longhorn-engine:${version} longhorn backup \
+              restore-to-file ${backup_url} \
+              --output-file /tmp/restore/${output_file} --output-format ${output_format} --backing-file ${backing_file}
+fi
```
Comment (innobead) [inline scripts/restore-backup-to-file.sh:96]: As per https://github.com/longhorn/longhorn-engine/blob/52ac3b6d0ca410957f53729c9aabbcb9e3bc007c/app/cmd/restore_to_file.go#L140-L140, `backing-file` is optional by including empty string, so you should not need to have if-else, but just one.

**Review by innobead — CHANGES_REQUESTED** [review]
Almost there, after resolving the rest of the comments, then it can be merged. 

Also, https://www.shellcheck.net/wiki/SC2086, always has a good habit to evaluate the variable as a string w/ double quote.

File: scripts/restore-backup-to-file.sh [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+   echo "Restore a Longhorn backup to a raw image or a qcow2 image."
+   echo ""
+   echo "      --aws-access-key         (Required) AWS credentials access key, e.g. AKIAIOSFODNN7EXAMPLE"
+   echo "      --aws-secret-access-key  (Required) AWS credentials access secret key, e.g. wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
+   echo "  -u, --backup-url             (Required) Backups URL, e.g. s3://backupbucket@us-east-1/backupstore?backup=backup-bd326da2c4414b02&volume=volumeexamplename"
+   echo "  -o, --output-file            (Required) Specify the output file name, which is saved in the /tmp/restore directory by default, e.g. volume.raw"
+   echo "  -f, --output-format          (Required) Output file foramt. Supported format: raw or qcow2."
+   echo "  -v, --version                (Required) Specify Lognhorn version to execute, e.g. v1.3.2"
+   echo "  -b, --backing-file           (Optional) If the backup is based on a backing file, it should provide the backing file, e.g. /tmp/bakingfile.qcow2"
+   echo "  -h, --help                   Usage message"
```
Comment (derekbit) [inline scripts/restore-backup-to-file.sh:69]: BTW, does it support restoring from NFS backup target?

Comment (innobead) [reply scripts/restore-backup-to-file.sh:69]: @weizhe0422 ?

Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh:69]: I think it can be retrieve from NFS, but I still try to verify it is workable.

Comment (innobead) [reply scripts/restore-backup-to-file.sh:69]: U should need to use privilege mode, because it's different from the s3 case.

Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh]: After fixing as per the comments, I confirmed that the backup destination can be NFS.
image

So I modified the logic to check the required parameters to only check for AWS related parameters when `s3` is set as the backup destination.  And also enable privilege mode when execute the docker container.

**Comment by weizhe0422** [comment]
Mounting NFS folder in docker container, after check the [Linux manual page](https://man7.org/linux/man-pages/man2/mount.2.html), there is need `SYS_ADMIN` capacity to mount filesystems. 

I think there are two ways to make mount filesystems successfully.
 1. Enabled all capabilities to the container: `docker run --privileged=true ...`
 2. Grant capabilities to perform administrator operations: `docker run --cap-add SYS_ADMIN --security-opt apparmor:unconfined...`

The [Docker document](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities) recommend to use the 2nd way. And using the 2nd way will make sure to mount keep kernel files ( e.g. `/sys/fs`, `/proc/fs`) as read-only.

Here is the result to benchmark using different parameters:
image

Ref.
 - https://serverfault.com/questions/824809/chrome-under-docker-cap-sys-admin-vs-privileged

File: scripts/restore-backup-to-file.sh [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+              --output-format "${output_format}" \
+              --backing-file "${backing_file}"
+else
+  docker run --cap-add SYS_ADMIN --security-opt apparmor:unconfined -v /tmp/restore:/tmp/restore \
+              longhornio/longhorn-engine:"${version}" longhorn backup \
+              restore-to-file ""${backup_url}"" \
+              --output-file "/tmp/restore/${output_file}" \
+              --output-format "${output_format}" \
+              --backing-file "${backing_file}"
+fi
```
Comment (derekbit) [inline scripts/restore-backup-to-file.sh:110]:
```suggestion
if [[ ! "${backup_url}" =~ ^[Ss]3 ]]; then
    DOCKER_CAP = "....."
    DOCKER_SEC = "....."
fi

docker run ${DOCKER_CAP} ${DOCKER_SEC} \
              -v /tmp/restore:/tmp/restore \
              longhornio/longhorn-engine:"${version}" longhorn backup \
              restore-to-file ""${backup_url}"" \
              --output-file "/tmp/restore/${output_file}" \
              --output-format "${output_format}" \
              --backing-file "${backing_file}"
fi
```

Probably can simplify the codes like this.

Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh]: Thank you. I will rewrite to simplify the code.

File: scripts/restore-backup-to-file.sh [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+  DOCKER_CAP_ARGS="--cap-add SYS_ADMIN --security-opt apparmor:unconfined"
+fi
+
+# Start restoring a backup to an image file. 
+docker run ${DOCKER_CAP_ARGS} -v /tmp/restore:/tmp/restore ${DOCKER_AWS_ARGS} \
+            longhornio/longhorn-engine:"${version}" longhorn backup \
+            restore-to-file ""${backup_url}"" \
+            --output-file "/tmp/restore/${output_file}" \
+            --output-format "${output_format}" \
+            --backing-file "${backing_file}"
```
Comment (derekbit) [inline scripts/restore-backup-to-file.sh:106]:
```suggestion
# Compose the docker arguments
if [[ "${backup_url}" =~ ^[Ss]3 ]]; then
            CUSTOMIZED_ARGS="-e AWS_ACCESS_KEY_ID="${aws_access_key}" -e AWS_SECRET_ACCESS_KEY="${aws_secret_access_key}" "
else
            CUSTOMIZED_ARGS="--cap-add SYS_ADMIN --security-opt apparmor:unconfined"
fi

# Start restoring a backup to an image file. 
docker run ${CUSTOMIZED_ARGS} -v /tmp/restore:/tmp/restore \
            longhornio/longhorn-engine:"${version}" longhorn backup \
            restore-to-file ""${backup_url}"" \
            --output-file "/tmp/restore/${output_file}" \
            --output-format "${output_format}" \
            --backing-file "${backing_file}"
```

File: scripts/restore-backup-to-file.sh [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+   echo "      --aws-access-key         (Optional) AWS credentials access key, e.g. AKIAIOSFODNN7EXAMPLE"
+   echo "      --aws-secret-access-key  (Optional) AWS credentials access secret key, e.g. wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
+   echo "  -b, --backing-file           (Optional) If the backup is based on a backing file, it should provide the backing file, e.g. /tmp/bakingfile.qcow2"
+   echo "  -h, --help                   Usage message"
+}
+
+if [[ "$help" ]]; then
+    usage
+    exit 0
+fi
```
Comment (derekbit) [inline scripts/restore-backup-to-file.sh:75]: See above comment.

**Review by derekbit — APPROVED** [review]

File: scripts/restore-backup-to-file.sh [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+  CUSTOMIZED_ARGS="--cap-add SYS_ADMIN --security-opt apparmor:unconfined"
+fi
+
+# Start restoring a backup to an image file. 
+docker run ${CUSTOMIZED_ARGS} -v /tmp/restore:/tmp/restore \
+            longhornio/longhorn-engine:"${version}" longhorn backup \
+            restore-to-file ""${backup_url}"" \
+            --output-file "/tmp/restore/${output_file}" \
+            --output-format "${output_format}" \
+            --backing-file "${backing_file}"
```
Comment (innobead) [inline scripts/restore-backup-to-file.sh:101]:
```suggestion
usage () {
   echo "USAGE: $0 --aws-access-key <your_aws_access_key>  \ "
   echo "          --aws-secret-access-key <your_aws_secret_access_key> \ "
   echo "          --backup-url s3://backupbucket@ap-northeast-1/backupstore?backup=<backup_name>&volume=<volume_name> \ "
   echo "          --target-directory /tmp/restore/volume.raw \ "
   echo "          --output-format raw \ "
   echo "          [--backing-file <backing_file_path>]"
   echo "Restore a Longhorn backup to a raw image or a qcow2 image."
   echo ""
   echo "  -u, --backup-url             (Required) Backups S3/NFS URL. e.g., s3://backupbucket@us-east-1/backupstore?backup=backup-bd326da2c4414b02&volume=volumeexamplename"
   echo "  -o, --output-file            (Required) Output file, e.g., /tmp/restore/volume.raw"
   echo "  -f, --output-format          (Required) Output file format, e.g., raw or qcow2"
   echo "  -v, --version                (Required) Longhorn version, e.g., v1.3.2"
   echo "      --aws-access-key         (Optional) AWS credentials access key"
   echo "      --aws-secret-access-key  (Optional) AWS credentials access secret key"
   echo "  -b, --backing-file           (Optional) backing image. e.g., /tmp/bakingfile.qcow2"
   echo "  -h, --help                   Usage message"
}

error_invalid_params() {
  usage
  exit 1
}

while [[ "$#" -gt 0 ]]; do
  key="$1"
  case $key in
    --aws-access-key)
    aws_access_key="$2"
    shift # past argument
    shift # past value
    ;;
    --aws-secret-access-key)
    aws_secret_access_key="$2"
    shift # past argument
    shift # past value
    ;;
    -u|--backup-url)
    backup_url="$2"
    shift # past argument
    shift # past value
    ;;
    -o|--output-file)
    output_file="$2"
    shift # past argument
    shift # past value
    ;;
    -f|--output-format)
    output_format="$2"
    shift # past argument
    shift # past value
    ;;
    -b|--backing-file)
    backing_file="$2"
    shift # past argument
    shift # past value
    ;;
    -v|--version)
    version="$2"
    shift # past argument
    shift # past value
    ;;
    -h|--help)
    usage
    exit 0
    shift
    ;;
    *)
    error_invalid_params
    ;;
    esac
done

# Check the required parameters exits
if [ -z "${backup_url}" ] || [ -z "${output_file}" ] || [ -z "${output_format}" ] || [ -z "${version}" ]; then
      error_invalid_params
fi

if [[ "${backup_url}" =~ ^[Ss]3 ]]; then
  if [ -z "${aws_access_key}" ] || [ -z "${aws_secret_access_key}" ]; then
      error_invalid_params
  fi
fi

# Compose the docker arguments
if [[ "${backup_url}" =~ ^[Ss]3 ]]; then
  CUSTOMIZED_ARGS="-e AWS_ACCESS_KEY_ID="${aws_access_key}" -e AWS_SECRET_ACCESS_KEY="${aws_secret_access_key}" "
else
  CUSTOMIZED_ARGS="--cap-add SYS_ADMIN --security-opt apparmor:unconfined"
fi

# Start restoring a backup to an image file. 
docker run ${CUSTOMIZED_ARGS} -v /tmp/restore:/tmp/restore \
            longhornio/longhorn-engine:"${version}" longhorn backup \
            restore-to-file ""${backup_url}"" \
            --output-file "/tmp/restore/${output_file}" \
            --output-format "${output_format}" \
            --backing-file "${backing_file}"
```

**Review by innobead — CHANGES_REQUESTED** [review]

File: scripts/restore-backup-to-file.sh (hunk 19 above)
Comment (weizhe0422) [reply scripts/restore-backup-to-file.sh]: Thank you. 

I have checked the suggested version and made some modifications.
